### PR TITLE
Add discovery to properly identify flannel cni

### LINKED ISSUE #1731: Auto-discovery of CNI is failing on K3s.
**What happened**: Auto-discovery of CNI is failing on K3s. It does not detect the CNI in the cluster and uses "generic" route-agent.

```
 ✓ Showing Gateways
NODE                            HA STATUS       SUMMARY                         
k3-subctl-w1                    active          All connections (1) are established

    Discovered network details via Submariner:
        Network plugin:  generic
        Service CIDRs:   [10.45.0.0/16]
        Cluster CIDRs:   [10.44.0.0/24]
 ✓ Showing Network details
```

**What you expected to happen**: It should be able to detect flannel CNI.

**How to reproduce it (as minimally and precisely as possible)**:
Follow the exact steps in URL https://submariner.io/getting-started/quickstart/k3s/.  below is exact steps to build two clusters

```
Cluster a:
POD_CIDR=10.44.0.0/16
SERVICE_CIDR=10.45.0.0/16
curl -sfL https://get.k3s.io | INSTALL_K3S_EXEC="--cluster-cidr $POD_CIDR --service-cidr $SERVICE_CIDR" sh -s -

sudo snap install yq

sudo su

cp /etc/rancher/k3s/k3s.yaml kubeconfig.cluster-a
export IP=$(hostname -I | awk '{print $1}')
yq -i eval \
'.clusters[].cluster.server |= sub("127.0.0.1", env(IP)) | .contexts[].name = "cluster-a" | .current-context = "cluster-a"' \
kubeconfig.cluster-a

Cluster b:

POD_CIDR=10.144.0.0/16
SERVICE_CIDR=10.145.0.0/16
curl -sfL https://get.k3s.io | INSTALL_K3S_EXEC="--cluster-cidr $POD_CIDR --service-cidr $SERVICE_CIDR" sh -s -

sudo snap install yq

sudo su

cp /etc/rancher/k3s/k3s.yaml kubeconfig.cluster-b
export IP=$(hostname -I | awk '{print $1}')
yq -i eval \
'.clusters[].cluster.server |= sub("127.0.0.1", env(IP)) | .contexts[].name = "cluster-b" | .current-context = "cluster-b"' \
kubeconfig.cluster-b
```

**Anything else we need to know?**:

**Environment**:
- Diagnose information (use `subctl diagnose all`):

```
walter@k3-subctl-m1:~$ subctl show all --kubeconfig kubeconfig.cluster-a
Cluster "default"
 ✓ Showing Connections
GATEWAY       CLUSTER    REMOTE IP     NAT  CABLE DRIVER  SUBNETS                       STATUS     RTT avg.  
k3-subctl-w2  cluster-b  192.168.1.93  no   libreswan     10.145.0.0/16, 10.144.0.0/24  connected            

 ✓ Showing Endpoints
CLUSTER ID                    ENDPOINT IP     PUBLIC IP       CABLE DRIVER        TYPE            
cluster-a                     192.168.1.92    58.182.182.152  libreswan           local           
cluster-b                     192.168.1.93    58.182.182.152  libreswan           remote          

 ✓ Showing Gateways
NODE                            HA STATUS       SUMMARY                         
k3-subctl-w1                    active          All connections (1) are established

    Discovered network details via Submariner:
        Network plugin:  generic
        Service CIDRs:   [10.45.0.0/16]
        Cluster CIDRs:   [10.44.0.0/24]
 ✓ Showing Network details

COMPONENT                       REPOSITORY                                            VERSION         
submariner                      quay.io/submariner                                    0.11.0          
submariner-operator             quay.io/submariner                                    0.11.0          
service-discovery               quay.io/submariner                                    0.11.0          
 ✓ Showing versions

walter@k3-subctl-m1:~$ subctl show all --kubeconfig kubeconfig.cluster-b
Cluster "default"
 ✓ Showing Connections
GATEWAY       CLUSTER    REMOTE IP     NAT  CABLE DRIVER  SUBNETS                     STATUS     RTT avg.  
k3-subctl-w1  cluster-a  192.168.1.92  no   libreswan     10.45.0.0/16, 10.44.0.0/24  connected            

 ✓ Showing Endpoints
CLUSTER ID                    ENDPOINT IP     PUBLIC IP       CABLE DRIVER        TYPE            
cluster-b                     192.168.1.93    58.182.182.152  libreswan           local           
cluster-a                     192.168.1.92    58.182.182.152  libreswan           remote          

 ✓ Showing Gateways
NODE                            HA STATUS       SUMMARY                         
k3-subctl-w2                    active          All connections (1) are established

    Discovered network details via Submariner:
        Network plugin:  generic
        Service CIDRs:   [10.145.0.0/16]
        Cluster CIDRs:   [10.144.0.0/24]
 ✓ Showing Network details

COMPONENT                       REPOSITORY                                            VERSION         
submariner                      quay.io/submariner                                    0.11.0          
submariner-operator             quay.io/submariner                                    0.11.0          
service-discovery               quay.io/submariner                                    0.11.0          
 ✓ Showing versions

walter@k3-subctl-m1:~$ 
```

- Gather information (use `subctl gather`):
[submariner-cluster-A-20211227124333.tar.gz](https://github.com/submariner-io/submariner/files/7784854/submariner-cluster-A-20211227124333.tar.gz)
[submariner-cluster-B-20211227124346.tar.gz](https://github.com/submariner-io/submariner/files/7784852/submariner-cluster-B-20211227124346.tar.gz)

- Cloud provider or hardware configuration:
VMWARE fusion to build K3S

- Install tools:
NA
- Others:
NA

### PR DESCRIPTION
**Pull request by nerdalert**
- this commit checks for a daemonset containing app=flannel
  and then extracts details from the associated configmap
  in the same fashion as the canal cni support.

Dependent on v0.12.0 with code freeze coming up

Fixes https://github.com/submariner-io/submariner-operator/issues/1731

Signed-off-by: Brent Salisbury <bsalisbu@redhat.com>

### CODE REVIEW

**Comment by nerdalert** [comment]
This is a WIP I wanted to throw out there in order to get proper discovery labeling for the flannel CNI. It is tested against microshift which uses flannel and was my primary driver as mentioned in #1731. 

@sridhargaddam and I discussed the primary goal is for proper labeling since the generic discovery seems to satisfy flannel the same way it does canal. I didnt see a clean way to label it without have a discovery function `flanelNet, err := discoverFlannelNetwork(clientSet)`. Just wanted to pause before I put too much time in it if this is something maintainers are interested in solving the issue this way (e.g. no tests and it doesnt fully discover the configmap from the daemonset `ConfigMapVolumeSource{LocalObjectReference:LocalObjectReference{Name:kube-flannel-cfg,}` in case the configmap isn't named `kube-flannel-cfg` like it is in Microshift).

I know the release is getting wrapped up and code freeze is around the corner, this obviously isn't a priority since the flannel plugin with Microshift is functioning with the generic discovery but at some point making flannel a first class citizen makes sense. Ty!

Output looks like so:

```shell
* ./broker-info.subm says broker is at: https://172.31.26.28:6443
* There are 1 labeled nodes in the cluster:
  - ip-172-31-26-28.ec2.internal
        Network plugin:  flannel
        Service CIDRs:   [10.43.0.0/16]
        Cluster CIDRs:   [10.42.0.0/16]
        Global CIDR:     242.0.0.0/16
 ✓ Discovering network details
 ✓ Retrieving Globalnet information from the Broker
 ✓ Validating Globalnet configuration
 ✓ Assigning Globalnet IPs
 ✓ Using pre-configured global CIDR 242.0.0.0/16
 ✓ Deploying the Submariner operator
 ✓ Created Lighthouse service accounts and roles
 ✓ Creating SA for cluster
 ✓ Deploying Submariner
 ✓ Submariner is up and running
```

**Comment by dfarrell07** [comment]
Thanks Brent! One minor suggestion regarding WIP PRs: https://submariner.io/development/code-review/#mark-work-in-progress-prs-as-drafts

**Comment by nerdalert** [comment]
Added a search through the daemonsets for a flannel set and then check for a volume containing an associated flannel configmap as mentioned above. This allows for a variance in the naming of the daemonsets/labels/configmaps as long as the names contain the flannel keyword.

**Comment by nerdalert** [comment]
bump

**Comment by nyechiel** [comment]
@nerdalert I know that you are no longer working on Submariner, but do you think you can have some bandwidth to complete this one?

**Comment by nerdalert** [comment]
@nyechiel of course! Will rebase and see if it still works this week.

**Comment by nyechiel** [comment]
> @nyechiel of course! Will rebase and see if it still works this week.

Thanks @nerdalert! @sridhargaddam FYI

**Comment by nerdalert** [comment]
@nyechiel @sridhargaddam looks like it just needed a rebase but will see if it passes build. Ty for looking at it frens.

**Comment by nyechiel** [comment]
@sridhargaddam appreciate your review here when you have some time

File: pkg/discovery/network/flannel.go
Comment (sridhargaddam) [inline pkg/discovery/network/flannel.go]:
```suggestion
		return nil, errors.WithMessage(err, "error listing the Daemonsets")
```

File: pkg/discovery/network/flannel_test.go
Comment (sridhargaddam) [inline pkg/discovery/network/flannel_test.go]: We are defining the list of CNI plugins in submariner repo and using it in operator. I'm fine to merge this PR and update it after the flannel constant is added in the list of CNIs - https://github.com/submariner-io/submariner/blob/85b5c9a7fe57632e75179c7ff55d59ece9cb2f98/pkg/cni/plugins.go#L22

**Review by sridhargaddam — COMMENTED** [review]
Sorry for the delay in reviewing the PR. 

Small nit, otherwise LGTM. Thanks @nerdalert

File: pkg/discovery/network/flannel_test.go
Comment (sridhargaddam) [reply pkg/discovery/network/flannel_test.go]: I pushed a PR in Submariner repo with the constant - https://github.com/submariner-io/submariner/pull/1941

Comment (nerdalert) [reply pkg/discovery/network/flannel_test.go]: Thanks @sridhargaddam! resolved the other comment.

**Review by tpantelis — APPROVED** [review]

**Review by dfarrell07 — APPROVED** [review]

**Comment by skitt** [comment]
@tpantelis can you fix the release notes so they pass MD linting?

**Review by skitt — APPROVED** [review]

**Pull request merged**
